src/centroiders.cpp

```diff
@@ -1,95 +1,199 @@
-            }
-        }
-        
-        result.push_back(closestIndex);
-    }
-
-    return result;
-}
-
-CentroidComparison StarCentroidsCompare(float threshold,
-                                    std::vector<Star> expected,
-                                    std::vector<Star> actual) {
-
-    CentroidComparison result;
-    // maps from indexes in each list to the closest centroid from other list
-    std::vector<int> expectedToActual, actualToExpected;
-
-    expectedToActual = FindClosestCentroids(threshold, expected, actual);
-    actualToExpected = FindClosestCentroids(threshold, actual, expected);
-
-    // any expected stars whose closest actual star does not refer to them are missing
-    for (int i = 0; i < (int)expectedToActual.size(); i++) {
-        if (expectedToActual[i] == -1 ||
-            actualToExpected[expectedToActual[i]] != i) {
-            result.numMissingStars++;
-        } else {
-            result.meanError += StarDistancePixels(expected[i], actual[expectedToActual[i]]);
-        }
-    }
-    result.meanError /= (expected.size() - result.numMissingStars);
-
-
-    // any actual star whose closest expected star does not refer to them is extra
-    for (int i = 0; i < (int)actual.size(); i++) {
-        if (actualToExpected[i] == -1 ||
-            expectedToActual[actualToExpected[i]] != i) {
-            result.numExtraStars++;
-        }
-    }
-
-    return result;
-}
-
-void CentroidComparison::Print() {
-    printf("Extra stars: %d\nMissing stars: %d\nMean error: %f\n",
-           numExtraStars, numMissingStars, meanError);
-}
-
-}
+#include "centroiders.hpp"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+
+#include <vector>
+#include <iostream>
+
+#include <unordered_map>
+
+#include <unordered_set>
+
+namespace lost {
+
+// DUMMY
+
+std::vector<Star> DummyCentroidAlgorithm::Go(unsigned char *image, int imageWidth, int imageHeight) const {
+    std::vector<Star> result;
+
+    for (int i = 0; i < numStars; i++) {
+        result.push_back(Star(rand() % imageWidth, rand() % imageHeight, 10.0));
+    }
+
+    return result;
+}
+
+//recursive helper here
+void cogHelper(float &yCoordMagSum,  
+    float &xCoordMagSum, 
+    int &magSum, 
+    int &xMin, 
+    int &xMax, 
+    int &yMin, 
+    int &yMax, 
+    int &cutoff, 
+    std::unordered_set<int> &checkedIndeces, int i, unsigned char *image, int imageWidth, int imageHeight) {
+    if (i >= 0 && i < imageWidth * imageHeight && image[i] >= cutoff && checkedIndeces.count(i) == 0) {
+        checkedIndeces.insert(i);
+        if (i % imageWidth > xMax) {
+            xMax = i % imageWidth;
+        } else if (i % imageWidth < xMin) {
+            xMin = i % imageWidth;
+        }
+        if (i / imageWidth > yMax) {
+            yMax = i / imageWidth;
+        } else if (i / imageWidth < yMin) {
+            yMin = i / imageWidth;
+        }
+        magSum += image[i];
+        xCoordMagSum += ((i % imageWidth) + 1) * image[i];

```

Comment:
What's the purpose of this +1?

Reply:
The equation for center of mass doesn't calculate correctly if there are 0 coordinates. because the image index starts at 0 I had to add one to the coordinate to get the right calculation for the center coordinates. i.e. If a star has pixels in the 0 x coordinate range, the coordinate * intensity at those points are always 0. Originally I thought we had to subtract 1.0 from the coordinate afterwards but it worked without the subtraction and I decided not to question it.

Reply:
I don't understand why the zero is a problem. When you are calculating the average of some numbers (and this works for weighted average too), it is ok if some of those numbers are zero. For example, the average of one two and three: (0+1+2)/3=1. It works because even the zero coordinates still contribute to magSum.

Reply:
Yeah you're right, I think when trying out an example I miscalculated

Reply:
Done - removed stray +1 from magSum calculations